File: dt241m-controller/internal/controller/discovery_test.go
package controller_test

import (
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/jacobgad/dt241m-controller/internal/discovery"
	"github.com/jacobgad/dt241m-controller/internal/dt241m"
	"github.com/jacobgad/dt241m-controller/internal/mqtt"
	"github.com/jacobgad/dt241m-controller/internal/testutil"
)

const (
	rxA = "aa:aa:aa:aa:aa:aa"
	rxB = "bb:bb:bb:bb:bb:bb"
)

func TestProbeFindsDevicesAmongUnresponsiveAddresses(t *testing.T) {
	net := testutil.NewNetwork()
	rx(testutil.RxFixtureMAC, "192.168.1.5", net)
	tx(testutil.TxFixtureMAC, "192.168.1.9", net)
	client := dt241m.NewHTTPClient(dt241m.Options{Transport: net, Timeout: 100 * time.Millisecond})
	hits := discovery.Probe(context.Background(), client, ips14(), discovery.Options{Concurrency: 4, Timeout: 100 * time.Millisecond})
	if len(hits) != 2 {
		t.Fatalf("hits %d", len(hits))
	}
	if len(net.Requests()) != 14 {
		t.Fatalf("expected every address probed, got %d", len(net.Requests()))
	}
}

func ips14() []string {
	out := make([]string, 0, 14)
	for i := 1; i <= 14; i++ {
		out = append(out, fmt.Sprintf("192.168.1.%d", i))
	}
	return out
}

func TestProbeContinuesPastHangingAddresses(t *testing.T) {
	net := testutil.NewNetwork()
	net.Unreachable = testutil.Hang
	rx(testutil.RxFixtureMAC, "192.168.1.14", net)
	client := dt241m.NewHTTPClient(dt241m.Options{Transport: net, Timeout: 30 * time.Millisecond})
	hits := discovery.Probe(context.Background(), client, ips14(), discovery.Options{Concurrency: 8, Timeout: 30 * time.Millisecond})
	if len(hits) != 1 || hits[0].IP != "192.168.1.14" {
		t.Fatalf("hits %+v", hits)
	}
}

func TestProbeRespectsConcurrencyLimit(t *testing.T) {
	net := testutil.NewNetwork()
	net.Unreachable = testutil.Hang
	client := dt241m.NewHTTPClient(dt241m.Options{Transport: net, Timeout: 20 * time.Millisecond})
	peak := 0
	discovery.Probe(context.Background(), client, ips14(), discovery.Options{
		Concurrency: 3, Timeout: 20 * time.Millisecond,
		OnProbeStart: func(_ string, active int) {
			if active > peak {
				peak = active
			}
		},
	})
	if peak != 3 {
		t.Fatalf("peak concurrency %d", peak)
	}
}

func TestDiscoveryRegistersClassifiesAndPublishes(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	tx(testutil.TxFixtureMAC, "192.168.1.9", h.net)
	h.discover(t)

	r, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	x, _ := h.ctrl.Registry.Get(testutil.TxFixtureMAC)
	if r.Role != dt241m.RoleReceiver || r.IP != "192.168.1.5" || x.Role != dt241m.RoleTransmitter || x.IP != "192.168.1.9" {
		t.Fatalf("rx %+v tx %+v", r, x)
	}
	configs := h.mqtt.DiscoveryConfigs()
	if _, ok := configs["homeassistant/number/dt241m_fc19286cd6d8/channel/config"]; !ok {
		t.Fatal("receiver number config missing")
	}
	if _, ok := configs["homeassistant/sensor/dt241m_fc19286cd291/channel/config"]; !ok {
		t.Fatal("transmitter sensor config missing")
	}
	if h.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).ChannelState) != "2" || h.mqtt.LastPayload(mqtt.ForDevice(testutil.TxFixtureMAC).ChannelState) != "3" {
		t.Fatal("channel states wrong")
	}
	if h.mqtt.LastPayload(mqtt.ControllerKnownState) != "2" || h.mqtt.LastPayload(mqtt.ControllerOnlineState) != "2" {
		t.Fatal("counts wrong")
	}
}

func TestSameMACAtNewIPUpdatesExistingDevice(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	h.discover(t)
	h.net.Move("192.168.1.5", "192.168.1.11")
	h.discover(t)
	if len(h.ctrl.Registry.All()) != 1 {
		t.Fatal("duplicate adapter")
	}
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if a.IP != "192.168.1.11" || h.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).IPState) != "192.168.1.11" {
		t.Fatalf("ip not updated: %+v", a)
	}
	if h.mqtt.LastPayload(mqtt.ControllerKnownState) != "1" {
		t.Fatal("known count wrong")
	}
}

func TestDifferentMACAtOldIPDoesNotMutateIdentity(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(rxA, "192.168.1.5", h.net)
	h.discover(t)
	h.net.Remove("192.168.1.5")
	rx(rxB, "192.168.1.5", h.net, func(o *testutil.DeviceOptions) { o.Overrides = map[string]any{"dev_name": "ER02_B"} })
	h.discover(t)

	a, _ := h.ctrl.Registry.Get(rxA)
	b, _ := h.ctrl.Registry.Get(rxB)
	if a.ID != "dt241m_aaaaaaaaaaaa" || *a.ReportedName != "ER02_286CD6D8" || a.Online {
		t.Fatalf("a mutated: %+v", a)
	}
	if b.ID != "dt241m_bbbbbbbbbbbb" || b.IP != "192.168.1.5" || !b.Online {
		t.Fatalf("b wrong: %+v", b)
	}
	if h.mqtt.LastPayload(mqtt.ForDevice(rxA).Availability) != "offline" || h.mqtt.LastPayload(mqtt.ForDevice(rxB).Availability) != "online" {
		t.Fatal("availability wrong")
	}
}

func TestRepeatedDiscoveryDoesNotDuplicateHADevice(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	for i := 0; i < 3; i++ {
		h.discover(t)
	}
	topics := map[string]bool{}
	uniqueIDs := map[string]bool{}
	for topic, cfg := range h.mqtt.DiscoveryConfigs() {
		if !strings.Contains(topic, "dt241m_fc19286cd6d8") {
			continue
		}
		topics[topic] = true
		uniqueIDs[cfg["unique_id"].(string)] = true
		ids := cfg["device"].(map[string]any)["identifiers"].([]any)
		if len(ids) != 1 || ids[0] != "dt241m:fc19286cd6d8" {
			t.Fatalf("identifiers %v", ids)
		}
	}
	expected := []string{
		"homeassistant/number/dt241m_fc19286cd6d8/channel/config",
		"homeassistant/sensor/dt241m_fc19286cd6d8/ip_address/config",
		"homeassistant/sensor/dt241m_fc19286cd6d8/role/config",
		"homeassistant/text/dt241m_fc19286cd6d8/name/config",
	}
	if len(topics) != len(expected) || len(uniqueIDs) != len(expected) {
		t.Fatalf("topics %v ids %v", topics, uniqueIDs)
	}
	for _, e := range expected {
		if !topics[e] {
			t.Fatalf("missing %s", e)
		}
	}
}

func TestNoOverlappingDiscovery(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	h.net.Unreachable = testutil.Hang
	done := make(chan struct{}, 2)
	go func() { h.ctrl.RunDiscovery(h.ctx, "a"); done <- struct{}{} }()
	eventually(t, h.ctrl.DiscoveryRunning, "discovery to start")
	go func() { h.ctrl.RunDiscovery(h.ctx, "b"); done <- struct{}{} }()
	<-done
	<-done
	if h.ctrl.DiscoveryRunCount() != 1 || h.ctrl.DiscoveryRunning() {
		t.Fatalf("runs %d running %v", h.ctrl.DiscoveryRunCount(), h.ctrl.DiscoveryRunning())
	}
}

func TestOnlyProbesConfiguredRange(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	h.discover(t)
	probed := map[string]bool{}
	for _, r := range h.net.Requests() {
		probed[r.IP] = true
	}
	if len(probed) != 14 || probed["192.168.1.0"] || probed["192.168.1.15"] {
		t.Fatalf("probed %v", probed)
	}
}

func TestPollingPublishesPhysicalChannelChangeWithoutWriting(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	d := rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	h.discover(t)
	d.SetReported(6)
	d.FrontPanel = 6
	h.ctrl.PollKnownDevices(h.ctx)
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if *a.Channel != 6 || h.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).ChannelState) != "6" {
		t.Fatal("channel not observed")
	}
	if len(h.net.WritesTo("192.168.1.5")) != 0 {
		t.Fatal("poll wrote to device")
	}
}

func TestPollingNeverReappliesChannels(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	d := rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	h.discover(t)
	if _, err := h.ctrl.RequestChannelChange(h.ctx, testutil.RxFixtureMAC, 4); err != nil {
		t.Fatal(err)
	}
	d.SetReported(1)
	h.ctrl.PollKnownDevices(h.ctx)
	h.ctrl.PollKnownDevices(h.ctx)
	if len(h.net.WritesTo("192.168.1.5")) != 1 {
		t.Fatal("poll reapplied channel")
	}
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if *a.Channel != 1 {
		t.Fatal("registry should track hardware")
	}
}

func TestVanishedDeviceIsRediscoveredAtNewIP(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	h.discover(t)
	runs := h.ctrl.DiscoveryRunCount()
	h.net.Move("192.168.1.5", "192.168.1.12")
	h.mqtt.Clear()
	h.ctrl.PollKnownDevices(h.ctx)

	availability := h.mqtt.PayloadsOn(mqtt.ForDevice(testutil.RxFixtureMAC).Availability)
	if len(availability) != 2 || availability[0] != "offline" || availability[1] != "online" {
		t.Fatalf("availability %v", availability)
	}
	if h.ctrl.DiscoveryRunCount() != runs+1 {
		t.Fatal("rediscovery not triggered")
	}
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if a.IP != "192.168.1.12" || len(h.ctrl.Registry.All()) != 1 {
		t.Fatalf("adapter %+v", a)
	}
}

func TestUnfoundDeviceStaysOffline(t *testing.T) {
	h := newHarness(t, harnessOptions{})
	rx(testutil.RxFixtureMAC, "192.168.1.5", h.net)
	h.discover(t)
	h.net.Remove("192.168.1.5")
	h.ctrl.PollKnownDevices(h.ctx)
	a, _ := h.ctrl.Registry.Get(testutil.RxFixtureMAC)
	if a.Online || h.mqtt.LastPayload(mqtt.ForDevice(testutil.RxFixtureMAC).Availability) != "offline" {
		t.Fatal("should be offline")
	}
	if h.mqtt.LastPayload(mqtt.ControllerOnlineState) != "0" || h.mqtt.LastPayload(mqtt.ControllerKnownState) != "1" {
		t.Fatal("counts wrong")
	}
}

File: dt241m-controller/src/tests/discovery.test.ts
import { describe, expect, it } from "vitest";
import { probeAddresses } from "../discovery.js";
import { createDt241mClient } from "../dt241m/client.js";
import { deviceTopics } from "../mqtt/topics.js";
import { RX_FIXTURE_MAC, TX_FIXTURE_MAC } from "./helpers/fixtures.js";
import { createHarness } from "./helpers/harness.js";
import { SimulatedDevice, SimulatedNetwork } from "./helpers/simulated-network.js";

const RX_A_MAC = "fc:19:28:6c:d6:d8";
const RX_B_MAC = "fc:19:28:6c:aa:01";

const ips = Array.from({ length: 14 }, (_, i) => `192.168.1.${i + 1}`);

describe("probeAddresses", () => {
  it("discovers one responsive device among unresponsive addresses", async () => {
    const net = new SimulatedNetwork();
    net.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const client = createDt241mClient({ fetchImpl: net.fetch });
    const hits = await probeAddresses(client, ips, { concurrency: 4, timeoutMs: 100 });
    expect(hits).toHaveLength(1);
    expect(hits[0]!.ip).toBe("192.168.1.5");
    expect(net.requests).toHaveLength(ips.length);
  });

  it("discovers multiple devices", async () => {
    const net = new SimulatedNetwork();
    net.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    net.place("192.168.1.9", new SimulatedDevice({ mac: TX_FIXTURE_MAC, fixture: "tx-info-initial-channel-3" }));
    const client = createDt241mClient({ fetchImpl: net.fetch });
    const hits = await probeAddresses(client, ips, { concurrency: 4, timeoutMs: 100 });
    expect(hits.map((h) => h.ip).sort()).toEqual(["192.168.1.5", "192.168.1.9"]);
  });

  it("continues past hanging addresses without aborting the scan", async () => {
    const net = new SimulatedNetwork();
    net.unreachableMode = "hang";
    net.place("192.168.1.14", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    const client = createDt241mClient({ fetchImpl: net.fetch });
    const hits = await probeAddresses(client, ips, { concurrency: 8, timeoutMs: 30 });
    expect(hits.map((h) => h.ip)).toEqual(["192.168.1.14"]);
  });

  it("respects the concurrency limit", async () => {
    const net = new SimulatedNetwork();
    net.unreachableMode = "hang";
    const client = createDt241mClient({ fetchImpl: net.fetch });
    let peak = 0;
    await probeAddresses(client, ips, {
      concurrency: 3,
      timeoutMs: 20,
      onProbeStart: (_ip, active) => {
        peak = Math.max(peak, active);
      }
    });
    expect(peak).toBe(3);
  });
});

describe("Controller discovery", () => {
  it("registers devices, classifies them, and publishes MQTT discovery", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    h.network.place("192.168.1.9", new SimulatedDevice({ mac: TX_FIXTURE_MAC, fixture: "tx-info-initial-channel-3" }));
    await h.controller.runDiscovery("test");

    const rx = h.controller.registry.get(RX_FIXTURE_MAC)!;
    const tx = h.controller.registry.get(TX_FIXTURE_MAC)!;
    expect(rx.role).toBe("receiver");
    expect(rx.ip).toBe("192.168.1.5");
    expect(tx.role).toBe("transmitter");
    expect(tx.ip).toBe("192.168.1.9");

    const configs = h.mqtt.discoveryConfigs();
    expect(configs.has("homeassistant/number/dt241m_fc19286cd6d8/channel/config")).toBe(true);
    expect(configs.has("homeassistant/sensor/dt241m_fc19286cd291/channel/config")).toBe(true);
    expect(h.mqtt.lastOn(deviceTopics(RX_FIXTURE_MAC).channelState)?.payload).toBe("2");
    expect(h.mqtt.lastOn(deviceTopics(TX_FIXTURE_MAC).channelState)?.payload).toBe("3");
    expect(h.mqtt.lastOn("dt241m/controller/known_devices/state")?.payload).toBe("2");
    expect(h.mqtt.lastOn("dt241m/controller/online_devices/state")?.payload).toBe("2");
  });

  it("updates the existing device when the same MAC appears at a new IP", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("first");
    h.network.move("192.168.1.5", "192.168.1.11");
    await h.controller.runDiscovery("second");

    expect(h.controller.registry.all()).toHaveLength(1);
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.ip).toBe("192.168.1.11");
    expect(h.mqtt.lastOn(deviceTopics(RX_FIXTURE_MAC).ipState)?.payload).toBe("192.168.1.11");
    expect(h.mqtt.lastOn("dt241m/controller/known_devices/state")?.payload).toBe("1");
  });

  it("does not mutate the old identity when its IP later answers with a different MAC", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_A_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("first");

    h.network.remove("192.168.1.5");
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_B_MAC, fixture: "rx-info-initial-channel-2", overrides: { dev_name: "ER02_B" } }));
    await h.controller.runDiscovery("second");

    const a = h.controller.registry.get(RX_A_MAC)!;
    const b = h.controller.registry.get(RX_B_MAC)!;
    expect(a.id).toBe("dt241m_fc19286cd6d8");
    expect(a.reportedName).toBe("ER02_286CD6D8");
    expect(a.online).toBe(false);
    expect(b.id).toBe("dt241m_fc19286caa01");
    expect(b.ip).toBe("192.168.1.5");
    expect(h.mqtt.lastOn(deviceTopics(RX_A_MAC).availability)?.payload).toBe("offline");
    expect(h.mqtt.lastOn(deviceTopics(RX_B_MAC).availability)?.payload).toBe("online");
  });

  it("repeated discovery does not duplicate the Home Assistant device identity", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("one");
    await h.controller.runDiscovery("two");
    await h.controller.runDiscovery("three");

    const configs = [...h.mqtt.discoveryConfigs().entries()].filter(([topic]) => topic.includes("dt241m_fc19286cd6d8"));
    const uniqueIds = new Set(configs.map(([, payload]) => payload["unique_id"]));
    expect(configs.map(([topic]) => topic).sort()).toEqual([
      "homeassistant/number/dt241m_fc19286cd6d8/channel/config",
      "homeassistant/sensor/dt241m_fc19286cd6d8/ip_address/config",
      "homeassistant/sensor/dt241m_fc19286cd6d8/role/config",
      "homeassistant/text/dt241m_fc19286cd6d8/name/config"
    ]);
    expect(uniqueIds.size).toBe(4);
    for (const [, payload] of configs) {
      expect((payload["device"] as { identifiers: string[] }).identifiers).toEqual(["dt241m:fc19286cd6d8"]);
    }
  });

  it("does not run overlapping discovery scans", async () => {
    const h = createHarness();
    h.network.unreachableMode = "hang";
    const first = h.controller.runDiscovery("a");
    const second = h.controller.runDiscovery("b");
    expect(h.controller.isDiscoveryRunning()).toBe(true);
    await Promise.all([first, second]);
    expect(h.controller.discoveryRunCount()).toBe(1);
    expect(h.controller.isDiscoveryRunning()).toBe(false);
  });

  it("only probes addresses inside the configured ranges", async () => {
    const h = createHarness();
    await h.controller.runDiscovery("bounded");
    const probed = new Set(h.network.requests.map((r) => r.ip));
    expect(probed.size).toBe(14);
    for (const ip of probed) expect(ip.startsWith("192.168.1.")).toBe(true);
    expect(probed.has("192.168.1.0")).toBe(false);
    expect(probed.has("192.168.1.15")).toBe(false);
  });
});

describe("Controller polling", () => {
  it("publishes a channel changed through physical controls", async () => {
    const h = createHarness();
    const device = h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("start");
    h.mqtt.clear();

    device.reportedChannel = 6;
    device.frontPanelChannel = 6;
    await h.controller.pollKnownDevices();

    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.channel).toBe(6);
    expect(h.mqtt.lastOn(deviceTopics(RX_FIXTURE_MAC).channelState)?.payload).toBe("6");
    expect(h.network.writesTo("192.168.1.5")).toHaveLength(0);
  });

  it("never reapplies channel settings while polling", async () => {
    const h = createHarness();
    const device = h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("start");
    await h.controller.requestChannelChange(RX_FIXTURE_MAC, 4);
    device.reportedChannel = 1;
    await h.controller.pollKnownDevices();
    await h.controller.pollKnownDevices();
    expect(h.network.writesTo("192.168.1.5")).toHaveLength(1);
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.channel).toBe(1);
  });

  it("marks a vanished device unavailable, rediscovers it at a new IP and keeps one HA device", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("start");
    const runsBefore = h.controller.discoveryRunCount();

    h.network.move("192.168.1.5", "192.168.1.12");
    h.mqtt.clear();
    await h.controller.pollKnownDevices();

    const availability = h.mqtt.messagesOn(deviceTopics(RX_FIXTURE_MAC).availability).map((m) => m.payload);
    expect(availability).toEqual(["offline", "online"]);
    expect(h.controller.discoveryRunCount()).toBe(runsBefore + 1);
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.ip).toBe("192.168.1.12");
    expect(h.controller.registry.all()).toHaveLength(1);
    expect(h.mqtt.lastOn(deviceTopics(RX_FIXTURE_MAC).ipState)?.payload).toBe("192.168.1.12");
  });

  it("leaves a device offline when it cannot be found anywhere", async () => {
    const h = createHarness();
    h.network.place("192.168.1.5", new SimulatedDevice({ mac: RX_FIXTURE_MAC, fixture: "rx-info-initial-channel-2" }));
    await h.controller.runDiscovery("start");
    h.network.remove("192.168.1.5");
    await h.controller.pollKnownDevices();
    expect(h.controller.registry.get(RX_FIXTURE_MAC)?.online).toBe(false);
    expect(h.mqtt.lastOn(deviceTopics(RX_FIXTURE_MAC).availability)?.payload).toBe("offline");
    expect(h.mqtt.lastOn("dt241m/controller/online_devices/state")?.payload).toBe("0");
    expect(h.mqtt.lastOn("dt241m/controller/known_devices/state")?.payload).toBe("1");
  });
});
